instance/models/server.py

```diff
@@ -76,54 +143,52 @@ def exclude_terminated(self):
         """
         Filter out terminated servers from the queryset
         """
-        return self.filter(~Q(status=Server.TERMINATED))
+        return self.filter(~Q(_status=Status.Terminated.state_id))
 
 
 class Server(ValidateModelMixin, TimeStampedModel):
     """
     A single server VM
     """
-    NEW = 'new'
-    STARTED = 'started'
-    ACTIVE = 'active'
-    BOOTED = 'booted'
-    PROVISIONING = 'provisioning'
-    REBOOTING = 'rebooting'
-    READY = 'ready'
-    LIVE = 'live'
-    STOPPING = 'stopping'
-    STOPPED = 'stopped'
-    TERMINATING = 'terminating'
-    TERMINATED = 'terminated'
-
-    STATUS_CHOICES = (
-        (NEW, 'New - Not yet loaded'),
-        (STARTED, 'Started - Running but not active yet'),
-        (ACTIVE, 'Active - Running but not booted yet'),
-        (BOOTED, 'Booted - Booted but not ready to be added to the application'),
-        (PROVISIONING, 'Provisioning - Provisioning is in progress'),
-        (REBOOTING, 'Rebooting - Reboot in progress, to apply changes from provisioning'),
-        (READY, 'Ready - Rebooted and ready to add to the application'),
-        (LIVE, 'Live - Is actively used in the application and/or accessed by users'),
-        (STOPPING, 'Stopping - Stopping temporarily'),
-        (STOPPED, 'Stopped - Stopped temporarily'),
-        (TERMINATING, 'Terminating - Stopping forever'),
-        (TERMINATED, 'Terminated - Stopped forever'),
+    Status = Status
+    status = ResourceStateDescriptor(state_classes=Status.states, default_state=Status.New, model_field_name='_status')
+    _status = models.CharField(
+        max_length=20,
+        default=status.default_state_class.state_id,
+        choices=status.model_field_choices,
+        db_index=True,
+        db_column='status',
     )
+    # State transitions:
+    _status_to_started = status.transition(from_states=Status.New, to_state=Status.Started)
+    _status_to_active = status.transition(from_states=Status.Started, to_state=Status.Active)
+    _status_to_booted = status.transition(from_states=Status.Active, to_state=Status.Booted)
+    _status_to_provisioning = status.transition(from_states=Status.Booted, to_state=Status.Provisioning)
+    _status_to_rebooting = status.transition(
+        from_states=(Status.Active, Status.Booted, Status.Ready, Status.Provisioning), to_state=Status.Rebooting
```

Comment:
@e-kolpakov Just to respond to your comments about the formal definition of an FSM:

You are correct that this does not match a formal description as you describe. I actually started implementing something more like that at first, but I found this approach to be more practical (for now) for a few reasons:
* There is no event loop in the app right now. There is a sort of "local event loop" on the server class with its `update_status` and `sleep_until_status` methods, and together they kind of approximate what you are describing. For example, you can think of the "mark_as_provisioning()" method as an event dispatcher which dispatches the PROVISIONING_STARTED event, and Booted state will catch that event and then transition to the "Provisioning" status. There is no other way to use a public method and get the `Server` instance into the "Provisioning" state besides sending that specific event when it is already in the Booted status. I'll admit this is a bit of a stretch though as it's not formally implemented using "events".
* We have to be careful with event handling since this app runs as multiple processes. For example, with a naive event implementation, if we dispatched a "server is ready for provisioning" event to a global event queue, then it's possible that either (A) multiple processes would catch that event and handle it by starting competing ansible runs, or (B) only one processes would catch that event (by design) and then the other processes' `Server.status` info would be out of date. These problems can be worked around, but I just want to point them out. Currently we work around them by having one worker "own" the process of provisioning and explicitly tell the server to update its status.
* The "list of states" vs. single state only is a pragmatic shortcut, to represent the fact that the server could fail to an error state or a terminated state at any time (due to a problem on the provider side), and saving us from having to define a separate transition on each state to the error/terminated states.